src/vga.js

```diff
@@ -1333,113 +2212,349 @@ VGAScreen.prototype.svga_register_read = function(n)
+                            byte3 >> i & 8;
+                }
+                break;
+
+            // Packed Shift Mode, aka Interleaved Shift Mode
+            // Video Modes 4h and 5h
+            case 0x20:
+                shift_loads[0] = (byte0 >> 6 & 0x3) | (byte2 >> 4 & 0xC);
+                shift_loads[1] = (byte0 >> 4 & 0x3) | (byte2 >> 2 & 0xC);
+                shift_loads[2] = (byte0 >> 2 & 0x3) | (byte2 >> 0 & 0xC);
+                shift_loads[3] = (byte0 >> 0 & 0x3) | (byte2 << 2 & 0xC);
+
+                shift_loads[4] = (byte1 >> 6 & 0x3) | (byte3 >> 4 & 0xC);
+                shift_loads[5] = (byte1 >> 4 & 0x3) | (byte3 >> 2 & 0xC);
+                shift_loads[6] = (byte1 >> 2 & 0x3) | (byte3 >> 0 & 0xC);
+                shift_loads[7] = (byte1 >> 0 & 0x3) | (byte3 << 2 & 0xC);
+                break;
+
+            // 256-Color Shift Mode
+            // Video Modes 13h and unchained 256 color
+            case 0x40:
+            case 0x60:
+                shift_loads[0] = byte0 >> 4 & 0xF;
+                shift_loads[1] = byte0 >> 0 & 0xF;
+                shift_loads[2] = byte1 >> 4 & 0xF;
+                shift_loads[3] = byte1 >> 0 & 0xF;
+                shift_loads[4] = byte2 >> 4 & 0xF;
+                shift_loads[5] = byte2 >> 0 & 0xF;
+                shift_loads[6] = byte3 >> 4 & 0xF;
+                shift_loads[7] = byte3 >> 0 & 0xF;
+                break;
+        }
+
+        if(pel_width)
+        {
+            // Assemble from two sets of 4 bits.
+            for(var i = 0, j = 0; i < 4; i++, pixel_addr++, j += 2)
+            {
+                var color256 = (shift_loads[j] << 4) | shift_loads[j + 1];
+                this.pixel_buffer[pixel_addr] = color256;
+                this.dac_color_use_set(pixel_addr, color256);
+            }
+        }
+        else
+        {
+            for(var i = 0; i < 8; i++, pixel_addr++)
+            {
+                this.pixel_buffer[pixel_addr] = shift_loads[i];
+                this.dac_color_use_set(pixel_addr, shift_loads[i]);
+            }
+        }
     }
+};
 
-    if(!this.dest_buffer)
+/**
+ * Transfers graphics from Pixel Buffer to Destination Image Buffer.
+ * The 4-bit/8-bit color indices in the Pixel Buffer are passed through
+ * the internal palette (dac_map) and the DAC palette (vga256_palette) to
+ * obtain the final 32 bit color that the Canvas API uses.
+ */
+VGAScreen.prototype.vga_redraw = function()
+{
+    var dac_changed = this.dac_changed_colors;
+    var dac_root = this.dac_color_use_root;
+    var dac_next = this.dac_color_use_next;
+
+    this.dac_color_has_changed.fill(0);
+
+    var color_use_disabled = this.dac_color_use_disabled;
+
+    // If half, or more than half the colors have changed, better to
+    // do a complete redraw than using the linked list.
+    color_use_disabled |= (this.attribute_mode & 0x40) && dac_changed.length > 127;
+    color_use_disabled |= !(this.attribute_mode & 0x40) && dac_changed.length > 7;
```

Comment:
Is the dac linked list solely an optimisation? Can you explain the optimisation to me? It seems to complicate the logic significantly, if it is solely an optimisation what applications benefit from it?

Reply:
Yes. The linked list was initially created for these reasons:

- There are many games (and probably many GUI programs out there) that relies on changing the DAC palette to create smooth transitions, fade in/out, and animations. Therefore, we need to do some kind of `redraw` at some point to update the destination buffer with the correct values (or else the screen remains blank/incorrect and unusable).
- I saw this comment:
https://github.com/copy/v86/blob/153c50f257b0ab8fa5d8707d9f4680f209d5a817/src/vga.js#L969-L974
- There was a significant amount of lag (about 1 second) when doing a complete redraw during the Windows 98 startup screen (e.g. notice the black gap at the top of the screen between 5 seconds and 6 seconds in [this video](https://drive.google.com/file/d/1FSt14j7ayKnaT7AS8lF_4S1i4ZX9GlEk/view)), so I thought that a complete redraw would be too expensive for smooth transitions.
- ~~I think~~ I'm obsessed with overengineering :stuck_out_tongue:

I thought it'll be quite inefficient to redraw the entire buffer if only a relatively small number of pixels needs to be updated. The linked list thing is effectively 256 independent linked list, each keeping track of the pixel addresses that use a given color.

During a redraw, we loop through the colors that have changed. For a given `color256`, we can get the first pixel address that uses that color using `dac_color_use_root[color256]`.

The subsequent pixel addresses that also uses that color is found using `dac_color_use_next[pixel_addr]`. We go through the list until we reach the end, with the value `VGA_UNUSED_COLOR` as the terminator.

![image](https://user-images.githubusercontent.com/2609018/34969620-85da614e-fad4-11e7-97b3-da81aee61ce4.png)

To keep the linked lists up to date, whenever we modify the `pixel_buffer`, we move the pixel address from the old list to the new list. We join the pixel's previous neighbour nodes together in the previous list (to remove the pixel), and insert the pixel into the front of the new list. This remove and insertion step takes constant time.

However, `complete_redraw`s are no longer that expensive. In non-svga modes, I've limited the redraw to only the pixel buffer (512KiB) instead of the entire svga buffer (defaulted to 8MiB)

https://github.com/ErnWong/v86/blob/be4e3ee2f54ac0b363071643c7fea3b906ef715c/src/vga.js#L983-L998

Yeah, I agree that this piece linked list adds quite a fair bit of complexity to the VGA controller, and I'm happy to remove it until we find a use case for it. In fact, I've been thinking about removing it, since `complete_redraw`s for small VGA resolutions are already pretty fast (I don't have quantitative figures though). The linked lists consumes about 4MiB of memory. Compare this with the 512MiB pixel buffer. Moreover, when the guest os / programs change the DAC palette, they usually change all of the pixels anyway.

I think that the only situation that will greatly benefit from this are with larger SVGA resolutions using 8bpp that do transitions for a small number of pixels.

What do you think?

Reply:
Thanks for the explanation. Yes, I think we should get rid of it for the reasons you stated (a complete redraw being fast enough anyway in non-svga modes). Could you still quickly get a number on how long complete redraw takes?

> I think that the only situation that will greatly benefit from this are with larger SVGA resolutions using 8bpp that do transitions for a small number of pixels.

I'm actually not aware of any OSes that use this mode, unless you explicitly configure them so. But even then, the usefulness of this mode over 16bpp svga is doubtful.

Reply:
> ...the usefulness of this mode over 16bpp svga is doubtful

Agree.

> Could you still quickly get a number on how long complete redraw takes?

Sure thing. Here's some very rough readings, which are absolute times from my laptop [after adding a few lines](https://gist.github.com/ErnWong/27be47ac4f8f35ea6b45a1b1c08291bd#file-vga-js-L2572).

- 8-bit color modes start at about 3ms/redraw and speeds up to about 2ms/redraw (probably due to v8's optimizations).

- 4-bit color modes start at about 6ms/redraw and quickly speeds up to about 3.5ms/redraw. It's probably slower because it has to go through both `dac_map` and `vga256_palette`

Are these the kind of numbers you're looking for?

Reply:
Yes, these numbers are good. Let's throw the linked list out.